File: core/klf/preview.go
package klf

import (
	"fmt"
	"sort"
	"strings"
)

// VocabularyEntry describes how a Run type renders across HTML,
// plain-text display, and CAT-tool chips. Mirrors VocabularyEntry in
// packages/format/src/vocabulary.ts.
type VocabularyEntry struct {
	Key         string
	Category    string
	HTML        HTMLRendering
	Display     TextRendering
	Chip        ChipRendering
	Constraints RunConstraints
}

// HTMLRendering is the template set applied when rendering a run as
// HTML. `{field}` placeholders expand against the run's context.
type HTMLRendering struct {
	Open        string
	Close       string
	Placeholder string
}

// TextRendering is the plain-text equivalent for terminals and for
// MT APIs that don't accept HTML.
type TextRendering struct {
	Open        string
	Close       string
	Placeholder string
}

// ChipRendering is the CAT-tool chip label + color.
type ChipRendering struct {
	Label string
	Color ChipColor
}

// ChipColor is the CSS color triple for a chip.
type ChipColor struct {
	Bg     string
	Border string
	Text   string
}

// JSXVocabulary mirrors the hand-curated default JSX vocabulary in
// packages/format/src/vocabulary.ts — the three span types the Go
// side needs to render Level-1 previews without loading the JSON
// vocabulary file. Kept in sync by a round-trip test against the
// embedded rich-jsx.json.
var JSXVocabulary = []VocabularyEntry{
	{
		Key:      "jsx:element",
		Category: "structure",
		HTML: HTMLRendering{
			Open:  `<{subType} data-neokapi-span="{id}">`,
			Close: `</{subType}>`,
		},
		Display: TextRendering{
			Open:  "<{subType}>",
			Close: "</{subType}>",
		},
		Chip: ChipRendering{
			Label: "{subType}",
			Color: ChipColor{Bg: "#e2e8f0", Border: "#94a3b8", Text: "#1e293b"},
		},
		Constraints: RunConstraints{Deletable: false, Cloneable: false, Reorderable: true},
	},
	{
		Key:      "jsx:var",
		Category: "variable",
		HTML: HTMLRendering{
			Placeholder: `<span class="neokapi-var" data-var="{equiv}" data-type="{subType}">{equiv}</span>`,
		},
		Display: TextRendering{
			Placeholder: "{{equiv}}",
		},
		Chip: ChipRendering{
			Label: "{equiv}",
			Color: ChipColor{Bg: "#dbeafe", Border: "#3b82f6", Text: "#1e40af"},
		},
		Constraints: RunConstraints{Deletable: false, Cloneable: true, Reorderable: true},
	},
	{
		Key:      "jsx:node",
		Category: "node",
		HTML: HTMLRendering{
			Placeholder: `<span class="neokapi-node" data-node="{id}" title="{data}">{equiv}</span>`,
		},
		Display: TextRendering{
			Placeholder: "«{equiv}»",
		},
		Chip: ChipRendering{
			Label: "⟨⟩",
			Color: ChipColor{Bg: "#fef3c7", Border: "#f59e0b", Text: "#92400e"},
		},
		Constraints: RunConstraints{Deletable: true, Cloneable: false, Reorderable: true},
	},
}

// VocabularyLookup resolves a run type string to its VocabularyEntry
// or returns nil if unknown. Matches the TS VocabularyLookup
// function type.
type VocabularyLookup func(typeKey string) *VocabularyEntry

// NewVocabularyLookup builds a lookup function from a flat slice.
func NewVocabularyLookup(entries []VocabularyEntry) VocabularyLookup {
	m := make(map[string]*VocabularyEntry, len(entries))
	for i := range entries {
		m[entries[i].Key] = &entries[i]
	}
	return func(key string) *VocabularyEntry {
		return m[key]
	}
}

// DefaultJSXVocabulary is a convenience lookup built from
// JSXVocabulary for tests, tools, and the Go-side preview reference
// implementation.
func DefaultJSXVocabulary() VocabularyLookup {
	return NewVocabularyLookup(JSXVocabulary)
}

// RenderBlockHTML wraps renderRuns in a <kat-block> marker — the
// same interactive wrapper neokapi's existing preview builders emit
// for every format. Matches renderBlockHtml in
// packages/format/src/preview.ts.
func RenderBlockHTML(b *Block, vocab VocabularyLookup) string {
	if vocab == nil {
		vocab = DefaultJSXVocabulary()
	}
	inner := RenderRuns(b.Source, vocab)
	return fmt.Sprintf(`<kat-block id=%q data-type=%q>%s</kat-block>`, b.ID, string(b.Type), inner)
}

// RenderRuns walks a run sequence in order and renders each run to
// its HTML form, recursing into plural / select forms. Mirrors
// renderRuns in packages/format/src/preview.ts.
func RenderRuns(runs []Run, vocab VocabularyLookup) string {
	var out strings.Builder
	for _, r := range runs {
		switch {
		case r.Text != nil:
			out.WriteString(escapeHTML(r.Text.Text))
		case r.Ph != nil:
			out.WriteString(renderEntry(vocab, r.Ph.Type, "placeholder", renderCtx{
				ID: r.Ph.ID, SubType: r.Ph.SubType, Data: r.Ph.Data, Equiv: r.Ph.Equiv,
			}))
		case r.PcOpen != nil:
			out.WriteString(renderEntry(vocab, r.PcOpen.Type, "open", renderCtx{
				ID: r.PcOpen.ID, SubType: r.PcOpen.SubType, Data: r.PcOpen.Data, Equiv: r.PcOpen.Equiv,
			}))
		case r.PcClose != nil:
			out.WriteString(renderEntry(vocab, r.PcClose.Type, "close", renderCtx{
				ID: r.PcClose.ID, SubType: r.PcClose.SubType, Data: r.PcClose.Data, Equiv: r.PcClose.Equiv,
			}))
		case r.Sub != nil:
			out.WriteString(fmt.Sprintf(`<span class="neokapi-sub" data-ref=%q>%s</span>`,
				escapeHTML(r.Sub.Ref), escapeHTML(r.Sub.Equiv)))
		case r.Plural != nil:
			out.WriteString(renderPlural(r.Plural, vocab))
		case r.Select != nil:
			out.WriteString(renderSelect(r.Select, vocab))
		}
	}
	return out.String()
}

type renderCtx struct {
	ID, SubType, Data, Equiv string
}

func renderEntry(vocab VocabularyLookup, typ, kind string, ctx renderCtx) string {
	entry := vocab(typ)
	if entry == nil {
		return fmt.Sprintf(`<span class="neokapi-unknown">%s</span>`, escapeHTML(ctx.Data))
	}
	var tmpl string
	switch kind {
	case "open":
		tmpl = entry.HTML.Open
	case "close":
		tmpl = entry.HTML.Close
	case "placeholder":
		tmpl = entry.HTML.Placeholder
	}
	return expandTemplate(tmpl, map[string]string{
		"id":      ctx.ID,
		"subType": ctx.SubType,
		"data":    escapeHTML(ctx.Data),
		"equiv":   escapeHTML(ctx.Equiv),
	})
}

// renderPlural emits forms in a deterministic order to keep output
// byte-stable for fixture tests. The TS reference implementation
// uses Object.entries order, which for our fixtures always matches
// ICU's canonical order (zero, one, two, few, many, other).
func renderPlural(p *PluralRun, vocab VocabularyLookup) string {
	var inner strings.Builder
	for _, form := range orderedPluralForms(p.Forms) {
		label := "plural:" + string(form)
		body := RenderRuns(p.Forms[form], vocab)
		fmt.Fprintf(&inner,
			`<div class="neokapi-plural-form" data-form=%q><span class="neokapi-plural-form-label">%s</span>%s</div>`,
			string(form), escapeHTML(label), body)
	}
	return fmt.Sprintf(`<span class="neokapi-plural" data-pivot=%q>%s</span>`,
		escapeHTML(p.Pivot), inner.String())
}

func renderSelect(s *SelectRun, vocab VocabularyLookup) string {
	var inner strings.Builder
	for _, key := range orderedSelectKeys(s.Cases) {
		label := "select:" + key
		body := RenderRuns(s.Cases[key], vocab)
		fmt.Fprintf(&inner,
			`<div class="neokapi-select-case" data-value=%q><span class="neokapi-select-case-label">%s</span>%s</div>`,
			key, escapeHTML(label), body)
	}
	return fmt.Sprintf(`<span class="neokapi-select" data-pivot=%q>%s</span>`,
		escapeHTML(s.Pivot), inner.String())
}

// pluralOrder is ICU's canonical plural form order; used to walk a
// PluralRun.Forms map deterministically.
var pluralOrder = []PluralForm{PluralZero, PluralOne, PluralTwo, PluralFew, PluralMany, PluralOther}

func orderedPluralForms(m map[PluralForm][]Run) []PluralForm {
	out := make([]PluralForm, 0, len(m))
	seen := make(map[PluralForm]bool, len(m))
	for _, f := range pluralOrder {
		if _, ok := m[f]; ok {
			out = append(out, f)
			seen[f] = true
		}
	}
	// Any non-standard keys get appended in sorted order after the
	// canonical ones so output stays stable across runs.
	extras := make([]string, 0)
	for k := range m {
		if !seen[k] {
			extras = append(extras, string(k))
		}
	}
	sort.Strings(extras)
	for _, k := range extras {
		out = append(out, PluralForm(k))
	}
	return out
}

func orderedSelectKeys(m map[string][]Run) []string {
	// 'other' sorts last to match ICU convention, then alpha for the rest.
	hasOther := false
	keys := make([]string, 0, len(m))
	for k := range m {
		if k == "other" {
			hasOther = true
			continue
		}
		keys = append(keys, k)
	}
	sort.Strings(keys)
	if hasOther {
		keys = append(keys, "other")
	}
	return keys
}

// expandTemplate replaces `{field}` placeholders in a vocabulary
// template with values from a context map. Matches expandTemplate
// in packages/format/src/vocabulary.ts.
func expandTemplate(tmpl string, ctx map[string]string) string {
	var out strings.Builder
	out.Grow(len(tmpl))
	for i := 0; i < len(tmpl); {
		if tmpl[i] != '{' {
			out.WriteByte(tmpl[i])
			i++
			continue
		}
		// Find matching '}'. Keys are `\w+` (ASCII-ish), so we only
		// need to scan to the next brace.
		end := -1
		for j := i + 1; j < len(tmpl); j++ {
			if tmpl[j] == '}' {
				end = j
				break
			}
		}
		if end < 0 {
			out.WriteByte(tmpl[i])
			i++
			continue
		}
		key := tmpl[i+1 : end]
		if !isWordKey(key) {
			// Not a placeholder, keep as-is.
			out.WriteByte(tmpl[i])
			i++
			continue
		}
		out.WriteString(ctx[key])
		i = end + 1
	}
	return out.String()
}

func isWordKey(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if !(r >= 'a' && r <= 'z') && !(r >= 'A' && r <= 'Z') &&
			!(r >= '0' && r <= '9') && r != '_' {
			return false
		}
	}
	return true
}

func escapeHTML(s string) string {
	var out strings.Builder
	out.Grow(len(s))
	for _, r := range s {
		switch r {
		case '&':
			out.WriteString("&amp;")
		case '<':
			out.WriteString("&lt;")
		case '>':
			out.WriteString("&gt;")
		case '"':
			out.WriteString("&quot;")
		default:
			out.WriteRune(r)
		}
	}
	return out.String()
}

File: packages/kapi-format/src/preview.ts
/**
 * Reference implementation of the Level-1 preview renderer.
 *
 * Walks a Block's runs and emits the `<kat-block>` HTML shape
 * neokapi's existing HTML and Markdown preview builders produce
 * for every format.
 *
 * The purpose of having this in @neokapi/kapi-format is:
 *   1. Validate that Blocks carry enough info to render a preview
 *      without reading the original source.
 *   2. Give @neokapi/i18n-react's test suite a way to assert "the
 *      extracted Block renders to exactly this HTML" without
 *      depending on neokapi's Go implementation.
 *   3. Give the Go-side `core/formats/jsx.PreviewBuilder`
 *      implementation a reference to port against, keeping the TS
 *      and Go sides in lockstep via shared fixtures.
 */

import type { Block, Run } from "./block.ts";
import { projectRuns, projectRunsText, type ModelRunSpec } from "./run-projection.ts";
import type { VocabularyEntry } from "./vocabulary.ts";
import { JSX_VOCABULARY, expandTemplate } from "./vocabulary.ts";

export type VocabularyLookup = (type: string) => VocabularyEntry | undefined;

export function createVocabulary(entries: VocabularyEntry[]): VocabularyLookup {
  const map = new Map(entries.map((e) => [e.key, e]));
  return (type) => map.get(type);
}

// ─── Run rendering ────────────────────────────────────────────────

/**
 * Render a run sequence to HTML by walking the array in order and
 * dispatching each run to its vocabulary entry. Recurses into
 * plural / select forms, which hold their own Run[] sub-sequences.
 */
export function renderRuns(runs: Run[], vocab: VocabularyLookup): string {
  return projectRunsText(runs, htmlSpec(vocab));
}

/** The HTML reading of a run sequence, declared kind by kind. */
function htmlSpec(vocab: VocabularyLookup): ModelRunSpec<string> {
  return {
    text: (r) => escapeHtml(r.text),
    ph: (r) =>
      renderEntry(vocab, r.ph.type, "placeholder", {
        id: r.ph.id,
        subType: r.ph.subType ?? "",
        data: r.ph.data,
        equiv: r.ph.equiv,
      }),
    pcOpen: (r) =>
      renderEntry(vocab, r.pcOpen.type, "open", {
        id: r.pcOpen.id,
        subType: r.pcOpen.subType ?? "",
        data: r.pcOpen.data,
        equiv: r.pcOpen.equiv,
      }),
    pcClose: (r) =>
      renderEntry(vocab, r.pcClose.type, "close", {
        id: r.pcClose.id,
        subType: r.pcClose.subType ?? "",
        data: r.pcClose.data,
        equiv: r.pcClose.equiv ?? "",
      }),
    sub: (r) =>
      `<span class="neokapi-sub" data-ref="${escapeHtml(r.sub.ref)}">${escapeHtml(r.sub.equiv)}</span>`,
    plural: (r) => renderPluralRun(r.plural, vocab),
    select: (r) => renderSelectRun(r.select, vocab),
    // A kind this build cannot render still occupies its place in the output,
    // as an empty element naming it — a preview that quietly omits content is
    // indistinguishable from content that was never there.
    fallback: (kind) => `<span class="neokapi-unknown" data-kind="${escapeHtml(kind)}"></span>`,
  };
}

// CLDR plural-form order, mirroring core/kbf.pluralOrder, so the Go and
// TypeScript renderers emit forms in the same sequence regardless of the
// order they were authored in.
const PLURAL_ORDER = ["zero", "one", "two", "few", "many", "other"];

function orderedPluralForms(forms: Partial<Record<string, Run[]>>): string[] {
  const present = new Set(Object.keys(forms));
  const ordered = PLURAL_ORDER.filter((f) => present.has(f));
  // Any non-standard keys follow the canonical ones, sorted, for stability.
  const extras = [...present].filter((f) => !PLURAL_ORDER.includes(f)).sort();
  return [...ordered, ...extras];
}

// Select-case order, mirroring core/kbf.orderedSelectKeys: the keys sorted
// alphabetically, with `other` last.
function orderedSelectKeys(cases: Record<string, Run[]>): string[] {
  const keys = Object.keys(cases)
    .filter((k) => k !== "other")
    .sort();
  if ("other" in cases) keys.push("other");
  return keys;
}

function renderPluralRun(
  plural: { pivot: string; forms: Partial<Record<string, Run[]>> },
  vocab: VocabularyLookup,
): string {
  const pivot = escapeHtml(plural.pivot);
  const formEntries = orderedPluralForms(plural.forms).map(
    (form) => [form, plural.forms[form] as Run[]] as [string, Run[]],
  );
  const inner = formEntries
    .map(([form, formRuns]) => {
      const label = `plural:${form}`;
      const body = renderRuns(formRuns, vocab);
      return (
        `<div class="neokapi-plural-form" data-form="${escapeHtml(form)}">` +
        `<span class="neokapi-plural-form-label">${escapeHtml(label)}</span>${body}</div>`
      );
    })
    .join("");
  return `<span class="neokapi-plural" data-pivot="${pivot}">${inner}</span>`;
}

function renderSelectRun(
  sel: { pivot: string; cases: Record<string, Run[]> },
  vocab: VocabularyLookup,
): string {
  const pivot = escapeHtml(sel.pivot);
  const inner = orderedSelectKeys(sel.cases)
    .map((value) => [value, sel.cases[value]] as [string, Run[]])
    .map(([value, caseRuns]) => {
      const label = `select:${value}`;
      const body = renderRuns(caseRuns, vocab);
      return (
        `<div class="neokapi-select-case" data-value="${escapeHtml(value)}">` +
        `<span class="neokapi-select-case-label">${escapeHtml(label)}</span>${body}</div>`
      );
    })
    .join("");
  return `<span class="neokapi-select" data-pivot="${pivot}">${inner}</span>`;
}

function renderEntry(
  vocab: VocabularyLookup,
  type: string,
  kind: "open" | "close" | "placeholder",
  context: { id: string; subType: string; data: string; equiv: string },
): string {
  const entry = vocab(type);
  if (!entry) {
    // Fallback: show the raw data. Keeps us resilient if a new
    // vocabulary key appears before the registry is updated.
    return `<span class="neokapi-unknown">${escapeHtml(context.data)}</span>`;
  }
  const template =
    kind === "open"
      ? entry.html.open
      : kind === "close"
        ? entry.html.close
        : entry.html.placeholder;
  return expandTemplate(template, {
    id: context.id,
    subType: context.subType,
    data: escapeHtml(context.data),
    equiv: escapeHtml(context.equiv),
  });
}

// ─── Block rendering ──────────────────────────────────────────────

/**
 * Render a whole Block wrapped in a `<kat-block>` marker — the
 * same interactive wrapper neokapi's existing preview builders
 * emit for every format. Block.source is a flat Run[]; plurals
 * and select constructs inside that sequence recurse naturally.
 */
export function renderBlockHtml(
  block: Block,
  vocab: VocabularyLookup = createVocabulary(JSX_VOCABULARY),
): string {
  const inner = renderRuns(block.source, vocab);
  return `<kat-block id="${block.id}" data-type="${block.type}">${inner}</kat-block>`;
}

function escapeHtml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

// ─── Validation ───────────────────────────────────────────────────

/**
 * Assert that a target run sequence preserves every required
 * source placeholder. A target may add content or restructure
 * text, but it cannot drop a required placeholder. Optional
 * placeholders (conditional JSX nodes) may be dropped.
 *
 * Returns a list of errors; empty array means valid.
 */
export function validateTargetAgainstSource(
  sourceBlock: Block,
  targetRuns: Run[],
): ValidationError[] {
  const errors: ValidationError[] = [];
  const targetNames = collectRunEquivs(targetRuns);

  for (const placeholder of sourceBlock.placeholders) {
    if (placeholder.optional) continue;
    if (!targetNames.has(placeholder.name)) {
      errors.push({
        blockId: sourceBlock.id,
        kind: "missing-placeholder",
        placeholder: placeholder.name,
        message: `target is missing required placeholder "${placeholder.name}"`,
      });
    }
  }

  return errors;
}

/**
 * The references a run sequence carries, as placeholder preservation counts
 * them: the `equiv` of ph / pcOpen / sub runs and the `pivot` of a plural or
 * select. Every branch is visited, not one — a reference in the `few` form
 * counts as much as one in `other`.
 *
 * Declared rather than walked, because a kind of run that carried a reference
 * and was forgotten here would not fail: the preservation check would simply
 * stop covering it, and a translation could drop that placeholder unnoticed.
 */
const REFERENCES: ModelRunSpec<string> = {
  text: { dropped: "a text run carries no reference" },
  ph: (r) => r.ph.equiv,
  pcOpen: (r) => r.pcOpen.equiv,
  // The closing half repeats its opening's reference; counting it twice would
  // make a mismatched pair look balanced.
  pcClose: { dropped: "counted once, on the opening half of the pair" },
  sub: (r) => r.sub.equiv,
  plural: {
    expand: (r) => [
      r.plural.pivot,
      ...Object.values(r.plural.forms).flatMap((form) => projectRuns(form ?? [], REFERENCES)),
    ],
  },
  select: {
    expand: (r) => [
      r.select.pivot,
      ...Object.values(r.select.cases).flatMap((c) => projectRuns(c ?? [], REFERENCES)),
    ],
  },
  // Filtered out by the caller: a kind this build cannot read carries no
  // reference it could compare, and the reporter has already said so.
  fallback: () => "",
};

/**
 * Walk a run sequence (including nested plural / select forms)
 * and collect every reference that counts toward placeholder
 * preservation: `equiv` of ph / pcOpen / sub runs, plus the
 * `pivot` of any plural / select construct encountered.
 */
function collectRunEquivs(runs: Run[]): Set<string> {
  return new Set(projectRuns(runs, REFERENCES).filter(Boolean));
}

export interface ValidationError {
  blockId: string;
  kind: "missing-placeholder" | "extra-placeholder" | "malformed-runs";
  placeholder?: string;
  message: string;
}
